gpio/__init__.py

```diff
@@ -0,0 +1,307 @@
+
+        # Add class to open pins
+        _open_pins[pin] = self
+
+    @staticmethod
+    def configured(pin):
+        """Get a configured GPIOPin instance where available.
+
+        Args:
+            pin (int): the GPIO pin to configured
+
+        Returns:
+            object: GPIOPin is configured, otherwise None
+        """
+        try:
+            # Implicitly convert str to int, ie: "1" -> 1
+            pin = int(pin)
+        except TypeError:
+            raise TypeError("pin must be an int")
+
+        return _open_pins.get(pin)
+
+    def get_direction(self):
+        '''Get the direction of pin
+
+        Returns:
+            str: "in" or "out"
+        '''
+        with open(os.path.join(self.root, 'direction'), FMODE) as f:
+            return f.read().strip()
+
+    def set_direction(self, mode):
+        '''Set the direction of pin
+
+        Args:
+            mode (str): "in" or "out"
+        '''
+        if mode not in (IN, OUT, LOW, HIGH):
+            raise ValueError("Unsupported pin mode {}".format(mode))
+
+        with open(os.path.join(self.root, 'direction'), FMODE) as f:
+            f.write(str(mode))
+            f.flush()
+
+    def set_active_low(self, active_low):
+        '''Set the direction of pin
+
+        Args:
+            mode (bool): True/False
+        '''
+        if not isinstance(active_low, bool):
+            raise ValueError("active_low must be True or False")
+
+        with open(os.path.join(self.root, 'active_low'), FMODE) as f:
+            f.write('1' if active_low else '0')
+            f.flush()
+
+    def read(self):
+        '''Read pin value'''
+        self.value.seek(0)
+        # Subtracting 48 converts an ASCII "0" or "1" to an int
+        # ord("0") == 48
+        try:
+            return self.value.read()[0] - 48
+        except TypeError:  # Handle Python 2.x
+            return int(self.value.read())
+
+    def write(self, value):
+        '''Write pin value'''
+        # Convert any truthy value explicitly to HIGH and vice versa
+        # this is about 3x faster than int(bool(value))
+        value = HIGH if value else LOW
+        # write as bytes, about 3x faster than string IO
+        self.value.write(b'1' if value else b'0')
+        # state.value.write(str(value).encode())  # Slow alternate for Python 2
+
+    def cleanup(self):
+        '''Clean up pin
+
+        Unexports the pin and deletes it from the open list.
+
+        '''
+        # Note: I have not put "cleanup" into the __del__ method since it's not
+        # always desireable to unexport pins at program exit.
+        # Additionally "open" can be deleted *before* the GPIOPin instance.
+        self.value.close()
+
+        if os.path.exists(self.root):
+            with _export_lock:
+                with open(GPIO_UNEXPORT, FMODE) as f:
+                    f.write(self.pin)
+                    f.flush()
+
+        del _open_pins[int(self.pin)]
+
+
+def cleanup(pin=None, assert_exists=False):

```

Comment:
this, setup and all the other functions should be replaced with class instance use. Don't duplicate code!

Just get the pin instance and call the correct method... they should be pretty minimal.

Reply:
Hacked back the weeds here and based them around `GPIOPin.configured(pin)`. Still getting 100kHz using the rpio compat approach.